File: go/internal/e2e/pre_mcp_tool_call_hook_e2e_test.go
package e2e

import (
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestPreMCPToolCallHookE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("rejects SDK preMcpToolCall callback hooks", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Hooks: &copilot.SessionHooks{
				OnPreMCPToolCall: func(copilot.PreMCPToolCallHookInput, copilot.HookInvocation) (*copilot.PreMCPToolCallHookOutput, error) {
					return &copilot.PreMCPToolCallHookOutput{MetaToUse: map[string]any{"injected": "by-hook"}}, nil
				},
			},
		})
		if err == nil {
			if session != nil {
				_ = session.Disconnect()
			}
			t.Fatal("expected SDK callback hooks to be rejected")
		}
		if !strings.Contains(err.Error(), unsupportedSDKHooksMessage) {
			t.Fatalf("expected unsupported hooks error, got %v", err)
		}
	})
}

File: nodejs/test/e2e/pre_mcp_tool_call_hook.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import type { SessionHooks } from "../../src/types.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

const UNSUPPORTED_SDK_HOOKS_MESSAGE = "SDK hook callbacks are no longer supported";

describe("pre_mcp_tool_call_hook", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    it("rejects SDK preMcpToolCall callback hooks", async () => {
        const hooks: SessionHooks = {
            onPreMcpToolCall: () => ({ metaToUse: { injected: "by-hook" } }),
        };

        await expect(
            client.createSession({
                onPermissionRequest: approveAll,
                hooks,
            })
        ).rejects.toThrow(UNSUPPORTED_SDK_HOOKS_MESSAGE);
    });
});
